# Admixture
We jointly inferred ancestry proportions and population structure with the program [ADMIXTURE](https://genome.cshlp.org/content/19/9/1655.full), version 1.3.0, asessing cross-validation error using 10-fold cross validation, evaluating models for 2 through 10 ancestral populations (K), and identifying the K that best fit the data as that which minimized the error rate obtained via 10-fold cross vlidation. ADMIXTURE takes as input a binary-compressed plink genotype dataset in bed format. We looped through analyses at levels of K with the following unix command line structure as an example for an implementation with genotypes that are not filtered on allele frequency or individual-level missingness:
```bash
for i in `seq 2 10`;do admixture -j4 --cv=10 oneperrad_trachylepis.bed $i > admixture_trachylepis_unfiltered_k${i}.log 2>&1;done
```
Cross-validation data for levels of K using this data set are in CrossValidationErrorTable.tsv, and we plot them in R with crossvalidationerrorplot.R, producing Supplementary Figure S4 panel G of our manuscript. We append population level ids to rows in the Q files output by ADMIXTURE, and that contain ancestry proportions for each of the ancestral populations, with AddIdInfoToAdmixtureOut.py, providing the Q file for the level of K of interest and the table mapping individuals to populations, sample_population_info.txt. We then generate separate pdf-formatted figures for levels of K 2 through 7, with admixture_plotcode.R. Plots are manually merged rather than more elegantly combining them with one script (sorry!). Plots on this data comprise Supplementary Data Figure S4 panels A through F of our manuscript. Supplementary Data Figure S5 in our manuscript performs similar analyses on genotypes filtered on allele frequency and individual missingness. 


